### PR TITLE
Improve forc-fmt to format std lib and core 

### PR DESCRIPTION
**Pull request by crodas**
## Description

Changes needed to build #5404 and #5405 

## Improvements
1. Use byte-offset instead of chars offset
2. Enhanced `asm` formatting

## Better output when tests fail

image

## Checklist

- [x] I have linked to any relevant issues.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

File: swayfmt/tests/mod.rs [hunk 1]
```diff
@@ -2310,3 +2310,380 @@ fn test_function() -> bool {
 "#,
     );
 }
+
+#[test]
+fn long_doc_break_new_line_bug() {
```
Comment (kayagokalp) [inline swayfmt/tests/mod.rs:2315]: maybe drop the bug in the test name for consistency

**Review by kayagokalp — APPROVED** [review]
nice, left a nit

**Review by kayagokalp — DISMISSED** [review]

File: examples/asm_return_tuple_pointer/src/main.sw [hunk 2]
```diff
@@ -2,11 +2,11 @@ script;
 
 fn adder(a: u64, b: u64, c: u64) -> (u64, u64) {
     let empty_tuple = (0u64, 0u64);
-    asm(output: empty_tuple, r1: a, r2: b, r3: c, r4, r5) {
-        add  r4 r1 r2; // add a & b and put the result in r4
-        add  r5 r2 r3; // add b & c and put the result in r5
-        sw   output r4 i0; // store the word in r4 in output + 0 words
-        sw   output r5 i1; // store the word in r5 in output + 1 word
+    asm( output: empty_tuple, r1: a, r2: b, r3: c, r4, r5 ) {
```
Comment (sdankel) [inline examples/asm_return_tuple_pointer/src/main.sw:5]: Should this be changed back?

**Review by sdankel — APPROVED** [review]
LGTM 🚢

File: docs/book/src/blockchain-development/access_control.md [hunk 3]
```diff
@@ -2,14 +2,18 @@
 
 <!-- This section should explain access control in Sway -->
 <!-- access_control:example:start -->
+
 Smart contracts require the ability to restrict access to and identify certain users or contracts. Unlike account-based blockchains, transactions in UTXO-based blockchains (i.e. Fuel) do not necessarily have a unique transaction sender. Additional logic is needed to handle this difference, and is provided by the standard library.
+
```
Comment (sdankel) [inline docs/book/src/blockchain-development/access_control.md:7]: These newlines aren't needed

File: swayfmt/src/utils/language/expr/mod.rs [hunk 4]
```diff
@@ -544,46 +550,110 @@ impl Format for Expr {
...
-                        write!(
-                            formatted_code,
-                            "\n{}{} ",
-                            formatter.indent_to_str()?,
-                            double_pipe_token.span().as_str()
-                        )?;
-                    }
-                    _ => {
-                        write!(formatted_code, " {} ", double_pipe_token.span().as_str())?;
+                let mut rhs_code = FormattedCode::new();
```
Comment (JoshuaBatty) [inline swayfmt/src/utils/language/expr/mod.rs:610]: See my comment on the same code here https://github.com/FuelLabs/sway/pull/5405#discussion_r1434583575

**Review by JoshuaBatty — DISMISSED** [review]
Looks good, thanks

**Review by JoshuaBatty — APPROVED** [review]

**Review by sdankel — APPROVED** [review]

**Review by kayagokalp — APPROVED** [review]

**Pull request merged**